pkg/app/stores/stores.go

```diff
@@ -74,38 +75,49 @@ type postgresStore struct {
 	r, w, admin *pgxpool.Pool
 	err         error
 	closed      bool
+
+	logger *zap.Logger
 }
 
 // Postgres returns shared postgres connection pools.
 // The pools can be used for read, write, and admin (alter table) operations.
 func (s *postgresStore) Postgres() (r, w, admin *pgxpool.Pool, _ error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	if s.r != nil || s.err != nil {
-		return s.r, s.w, s.admin, s.err
-	}
 
 	fail := func(err error) (_, _, _ *pgxpool.Pool, _ error) {
-		s.err = err
 		return nil, nil, nil, err
 	}
+
+	// during shutdown, a caller may sporadically try to get a store
+	// after close has been called
+	if errors.Is(s.err, ErrStoreClosed) {
+		return fail(s.err)
+	}
+
+	if s.r != nil {
+		return s.r, s.w, s.admin, nil
+	}
+
 	if s.cfg == nil {
 		return fail(fmt.Errorf("%w: postgres", ErrStoreNotConfigured))
 	}
 
 	// todo: support r, w, and admin pools
 	pool, err := pgxutil.Connect(context.Background(), s.cfg.ConnectConfig)
 	if err != nil {
+		s.logger.Error("failed to connect to postgres", zap.Error(err))
```

Comment:
I don't think we need this log line now that we're returning the connection error.